File: android/app/src/main/java/com/lanchat/core/Transfer.kt
package com.lanchat.core

import kotlinx.coroutines.*
import org.json.JSONArray
import org.json.JSONObject
import java.io.*
import java.net.ServerSocket
import java.net.Socket
import java.security.MessageDigest
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

const val PARALLEL_STREAMS = 4

data class FileMeta(
    val name: String,
    val size: Long,
    val mime: String,
    val sha256: String,
    val isApp: Boolean = false,
    val appLabel: String? = null,
    val appPackage: String? = null,
)

data class FileToSend(val path: String, val name: String, val mime: String, val isApp: Boolean = false, val appLabel: String? = null, val appPackage: String? = null)

private fun byteRanges(size: Long, n: Int): List<LongArray> {
    val base = size / n
    var start = 0L
    val ranges = mutableListOf<LongArray>()
    for (i in 0 until n) {
        val len = if (i == n - 1) size - start else base
        ranges.add(longArrayOf(start, start + len - 1)) // inclusive
        start += len
    }
    return ranges
}

private fun sha256Of(file: File): String {
    val digest = MessageDigest.getInstance("SHA-256")
    file.inputStream().use { input ->
        val buf = ByteArray(64 * 1024)
        while (true) {
            val n = input.read(buf)
            if (n < 0) break
            digest.update(buf, 0, n)
        }
    }
    return digest.digest().joinToString("") { "%02x".format(it) }
}

private fun uuidToBytes(uuid: String): ByteArray {
    val hex = uuid.replace("-", "")
    return ByteArray(16) { i -> hex.substring(i * 2, i * 2 + 2).toInt(16).toByte() }
}
private fun bytesToUuid(b: ByteArray): String {
    val h = b.joinToString("") { "%02x".format(it) }
    return "${h.substring(0,8)}-${h.substring(8,12)}-${h.substring(12,16)}-${h.substring(16,20)}-${h.substring(20)}"
}

/** Receiving side: accepts offers, and on acceptance receives the parallel data streams. */
class TransferServer(
    private val self: SelfInfo,
    private val trustStore: TrustStore,
    private val receivedDir: File,
) {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var controlServer: ServerSocket? = null
    private val dataServers = mutableListOf<ServerSocket>()

    private data class ActiveTransfer(
        val files: List<FileMeta>,
        val destFiles: List<File>,
        val sessionKey: ByteArray,
        val controlOut: DataOutputStream,
        val randomAccessFiles: List<RandomAccessFile>,
        val streamsDoneByFile: IntArray,
        var filesDone: Int = 0,
    )
    private val transfers = ConcurrentHashMap<String, ActiveTransfer>()

    var onOffer: ((transferId: String, from: String, files: List<FileMeta>) -> Unit)? = null
    var onComplete: ((transferId: String, status: String, destPaths: List<String>) -> Unit)? = null

    fun start() {
        receivedDir.mkdirs()
        controlServer = ServerSocket(self.transferPort)
        scope.launch { acceptLoop(controlServer!!) { handleControl(it) } }
        for (i in 1..PARALLEL_STREAMS) {
            val srv = ServerSocket(self.transferPort + i)
            dataServers.add(srv)
            scope.launch { acceptLoop(srv) { handleData(it, i - 1) } }
        }
    }

    private suspend fun acceptLoop(server: ServerSocket, handler: suspend (Socket) -> Unit) {
        while (currentCoroutineContext().isActive) {
            val socket = try { server.accept() } catch (_: Exception) { null } ?: break
            scope.launch { handler(socket) }
        }
    }

    private fun handleControl(socket: Socket) {
        val input = DataInputStream(socket.getInputStream())
        val output = DataOutputStream(socket.getOutputStream())

        val helloLen = input.readInt()
        val helloBytes = ByteArray(helloLen).also { input.readFully(it) }
        val hello = JSONObject(String(helloBytes))
        val peerDeviceId = hello.getString("deviceId")
        val sessionKey = trustStore.sessionKeyFor(peerDeviceId) ?: run { socket.close(); return }

        try {
            while (true) {
                val len = input.readInt()
                val frame = ByteArray(len).also { input.readFully(it) }
                val plaintext = CryptoUtil.decrypt(sessionKey, frame)
                val msg = JSONObject(String(plaintext))
                if (msg.getString("type") == "offer") {
                    val transferId = msg.getString("transferId")
                    val filesJson = msg.getJSONArray("files")
                    val files = (0 until filesJson.length()).map { i ->
                        val f = filesJson.getJSONObject(i)
                        FileMeta(
                            f.getString("name"), f.getLong("size"), f.getString("mime"), f.getString("sha256"),
                            f.optBoolean("isApp", false), f.optString("appLabel", null), f.optString("appPackage", null),
                        )
                    }
                    val destFiles = files.map { File(receivedDir, it.name) }
                    val rafs = destFiles.map { RandomAccessFile(it, "rw").apply { setLength(0) } }
                    transfers[transferId] = ActiveTransfer(files, destFiles, sessionKey, output, rafs, IntArray(files.size))
                    onOffer?.invoke(transferId, peerDeviceId, files)
                }
            }
        } catch (_: Exception) {
            // control channel closed after transfer completes — expected
        }
    }

    /** Call from UI once user taps Accept. */
    fun acceptTransfer(transferId: String) {
        val t = transfers[transferId] ?: return
        val plaintext = JSONObject().put("type", "accept").put("transferId", transferId).toString().toByteArray()
        val frame = CryptoUtil.encrypt(t.sessionKey, plaintext)
        synchronized(t.controlOut) {
            t.controlOut.writeInt(frame.size)
            t.controlOut.write(frame)
        }
    }

    private fun handleData(socket: Socket, expectedStreamIndex: Int) {
        val input = DataInputStream(socket.getInputStream())
        val header = ByteArray(20)
        input.readFully(header)
        val transferId = bytesToUuid(header.copyOfRange(0, 16))
        val fileIndex = ((header[16].toInt() and 0xFF) shl 8) or (header[17].toInt() and 0xFF)
        val streamIndex = ((header[18].toInt() and 0xFF) shl 8) or (header[19].toInt() and 0xFF)

        val t = transfers[transferId] ?: run { socket.close(); return }
        val range = byteRanges(t.files[fileIndex].size, PARALLEL_STREAMS)[streamIndex]
        val raf = t.randomAccessFiles[fileIndex]
        var pos = range[0]

        val buf = ByteArray(64 * 1024)
        try {
            while (true) {
                val n = input.read(buf)
                if (n < 0) break
                synchronized(raf) {
                    raf.seek(pos)
                    raf.write(buf, 0, n)
                }
                pos += n
            }
        } catch (_: Exception) { /* stream ended */ }

        val done = synchronized(t) {
            t.streamsDoneByFile[fileIndex] += 1
            t.streamsDoneByFile[fileIndex] == PARALLEL_STREAMS
        }
        if (done) {
            raf.close()
            checkFileAndMaybeFinish(transferId, fileIndex)
        }
    }

    private fun checkFileAndMaybeFinish(transferId: String, fileIndex: Int) {
        val t = transfers[transferId] ?: return
        val ok = sha256Of(t.destFiles[fileIndex]) == t.files[fileIndex].sha256
        val allDone = synchronized(t) {
            t.filesDone += 1
            t.filesDone == t.files.size
        }
        if (allDone) {
            val status = if (ok) "ok" else "checksum_failed"
            val plaintext = JSONObject().put("type", "complete").put("transferId", transferId).put("status", status).toString().toByteArray()
            val frame = CryptoUtil.encrypt(t.sessionKey, plaintext)
            synchronized(t.controlOut) {
                t.controlOut.writeInt(frame.size)
                t.controlOut.write(frame)
            }
            onComplete?.invoke(transferId, status, t.destFiles.map { it.absolutePath })
            transfers.remove(transferId)
        }
    }

    fun stop() {
        scope.cancel()
        controlServer?.close()
        dataServers.forEach { it.close() }
    }
}

/** Sending side: offers files (or an installed app's APK), then streams them once accepted. */
class TransferClient(private val self: SelfInfo, private val trustStore: TrustStore) {

    suspend fun sendFiles(peer: PeerInfo, files: List<FileToSend>): String = withContext(Dispatchers.IO) {
        val sessionKey = trustStore.sessionKeyFor(peer.deviceId)
            ?: throw IllegalStateException("Cannot send: peer is not paired yet")

        val transferId = UUID.randomUUID().toString()
        val metas = files.map {
            FileMeta(it.name, File(it.path).length(), it.mime, sha256Of(File(it.path)), it.isApp, it.appLabel, it.appPackage)
        }

        val controlSocket = Socket(peer.ip, peer.transferPort)
        val controlOut = DataOutputStream(controlSocket.getOutputStream())
        val controlIn = DataInputStream(controlSocket.getInputStream())

        val hello = JSONObject().put("type", "hello").put("deviceId", self.deviceId).toString().toByteArray()
        controlOut.writeInt(hello.size); controlOut.write(hello)

        val filesJson = JSONArray()
        metas.forEach { m ->
            filesJson.put(JSONObject().apply {
                put("name", m.name); put("size", m.size); put("mime", m.mime); put("sha256", m.sha256)
                put("isApp", m.isApp); m.appLabel?.let { put("appLabel", it) }; m.appPackage?.let { put("appPackage", it) }
            })
        }
        val offerPlain = JSONObject().put("type", "offer").put("transferId", transferId).put("files", filesJson).toString().toByteArray()
        val offerFrame = CryptoUtil.encrypt(sessionKey, offerPlain)
        controlOut.writeInt(offerFrame.size); controlOut.write(offerFrame)

        // Wait for accept, then stream, then wait for complete.
        var status = "unknown"
        while (true) {
            val len = controlIn.readInt()
            val frame = ByteArray(len).also { controlIn.readFully(it) }
            val msg = JSONObject(String(CryptoUtil.decrypt(sessionKey, frame)))
            when (msg.getString("type")) {
                "accept" -> streamAllFiles(peer, transferId, files, metas)
                "complete" -> { status = msg.getString("status"); controlSocket.close(); break }
            }
        }
        status
    }

    private suspend fun streamAllFiles(peer: PeerInfo, transferId: String, files: List<FileToSend>, metas: List<FileMeta>) = coroutineScope {
        for (fileIndex in files.indices) {
            val ranges = byteRanges(metas[fileIndex].size, PARALLEL_STREAMS)
            ranges.forEachIndexed { streamIndex, range ->
                launch(Dispatchers.IO) {
                    val socket = Socket(peer.ip, peer.transferPort + streamIndex + 1)
                    val out = socket.getOutputStream()
                    val header = ByteArray(20)
                    uuidToBytes(transferId).copyInto(header, 0)
                    header[16] = ((fileIndex shr 8) and 0xFF).toByte(); header[17] = (fileIndex and 0xFF).toByte()
                    header[18] = ((streamIndex shr 8) and 0xFF).toByte(); header[19] = (streamIndex and 0xFF).toByte()
                    out.write(header)

                    RandomAccessFile(files[fileIndex].path, "r").use { raf ->
                        raf.seek(range[0])
                        var remaining = range[1] - range[0] + 1
                        val buf = ByteArray(64 * 1024)
                        while (remaining > 0) {
                            val n = raf.read(buf, 0, minOf(buf.size.toLong(), remaining).toInt())
                            if (n < 0) break
                            out.write(buf, 0, n)
                            remaining -= n
                        }
                    }
                    socket.close()
                }
            }
        }
    }
}

File: desktop/src/core/transfer.js
'use strict';
/**
 * File/app transfer ("Xender" capability).
 *
 * One small encrypted control connection (transferPort) negotiates the
 * transfer (offer -> accept -> complete). The actual bytes move over N
 * parallel *raw* TCP sockets (transferPort+1 .. transferPort+N), each one
 * responsible for a contiguous byte range of the file — this is what gives
 * LAN transfers real throughput instead of single-stream TCP being the
 * bottleneck.
 */
const net = require('net');
const fs = require('fs');
const crypto = require('crypto');
const EventEmitter = require('events');
const cryptoCore = require('./crypto');

const PARALLEL_STREAMS = 4;

function uuidToBuf16(uuid) {
  return Buffer.from(uuid.replace(/-/g, ''), 'hex');
}
function bufToUuid(buf) {
  const h = buf.toString('hex');
  return `${h.slice(0, 8)}-${h.slice(8, 12)}-${h.slice(12, 16)}-${h.slice(16, 20)}-${h.slice(20)}`;
}

function writeControlFrame(socket, sessionKey, obj) {
  const plaintext = Buffer.from(JSON.stringify(obj), 'utf8');
  const enc = cryptoCore.encrypt(sessionKey, plaintext);
  const header = Buffer.alloc(4);
  header.writeUInt32BE(enc.length, 0);
  socket.write(Buffer.concat([header, enc]));
}

function controlFrameReader(socket, sessionKey, onMessage) {
  let buffer = Buffer.alloc(0);
  socket.on('data', (chunk) => {
    buffer = Buffer.concat([buffer, chunk]);
    while (buffer.length >= 4) {
      const len = buffer.readUInt32BE(0);
      if (buffer.length < 4 + len) break;
      const frame = buffer.subarray(4, 4 + len);
      buffer = buffer.subarray(4 + len);
      const plaintext = cryptoCore.decrypt(sessionKey, frame);
      onMessage(JSON.parse(plaintext.toString('utf8')));
    }
  });
}

function byteRanges(size, n) {
  const ranges = [];
  const base = Math.floor(size / n);
  let start = 0;
  for (let i = 0; i < n; i++) {
    const len = i === n - 1 ? size - start : base;
    ranges.push({ start, end: start + len - 1 }); // inclusive, fs.createReadStream style
    start += len;
  }
  return ranges;
}

function sha256File(filePath) {
  return new Promise((resolve, reject) => {
    const hash = crypto.createHash('sha256');
    const stream = fs.createReadStream(filePath);
    stream.on('data', (d) => hash.update(d));
    stream.on('end', () => resolve(hash.digest('hex')));
    stream.on('error', reject);
  });
}

class TransferServer extends EventEmitter {
  constructor(self, trustStore, receivedDir) {
    super();
    this.self = self;
    this.trustStore = trustStore;
    this.receivedDir = receivedDir;
    this.transfers = new Map(); // transferId -> { files, destPaths, streamsRemaining, sessionKey, controlSocket }
    this.controlServer = null;
    this.dataServers = [];
  }

  start() {
    fs.mkdirSync(this.receivedDir, { recursive: true });

    this.controlServer = net.createServer((socket) => this._handleControl(socket));
    this.controlServer.listen(this.self.transferPort);

    for (let i = 1; i <= PARALLEL_STREAMS; i++) {
      const srv = net.createServer((socket) => this._handleData(socket));
      srv.listen(this.self.transferPort + i);
      this.dataServers.push(srv);
    }
  }

  _handleControl(socket) {
    let peerDeviceId = null;
    let sessionKey = null;
    let buffer = Buffer.alloc(0);

    socket.once('data', (chunk) => {
      // first plaintext frame: hello
      const len = chunk.readUInt32BE(0);
      const helloFrame = chunk.subarray(4, 4 + len);
      const hello = JSON.parse(helloFrame.toString('utf8'));
      peerDeviceId = hello.deviceId;
      sessionKey = this.trustStore.sessionKeyFor(peerDeviceId);
      if (!sessionKey) { socket.destroy(); return; }

      buffer = chunk.subarray(4 + len);
      const processRemaining = () => {
        while (buffer.length >= 4) {
          const l = buffer.readUInt32BE(0);
          if (buffer.length < 4 + l) break;
          const frame = buffer.subarray(4, 4 + l);
          buffer = buffer.subarray(4 + l);
          const plaintext = cryptoCore.decrypt(sessionKey, frame);
          this._onControlMessage(peerDeviceId, sessionKey, socket, JSON.parse(plaintext.toString('utf8')));
        }
      };
      processRemaining();
      socket.on('data', (c) => { buffer = Buffer.concat([buffer, c]); processRemaining(); });
    });
  }

  _onControlMessage(peerDeviceId, sessionKey, socket, msg) {
    if (msg.type === 'offer') {
      const destPaths = msg.files.map((f) => `${this.receivedDir}/${f.name}`);
      this.transfers.set(msg.transferId, {
        files: msg.files,
        destPaths,
        sessionKey,
        controlSocket: socket,
        fds: destPaths.map((p) => fs.openSync(p, 'w')),
      });
      this.emit('offer', { transferId: msg.transferId, from: peerDeviceId, files: msg.files });
    }
  }

  /** Call this from the UI once the user taps "Accept". */
  acceptTransfer(transferId) {
    const t = this.transfers.get(transferId);
    if (!t) throw new Error('Unknown transfer');
    writeControlFrame(t.controlSocket, t.sessionKey, { type: 'accept', transferId });
  }

  _handleData(socket) {
    socket.once('data', (header) => {
      const transferId = bufToUuid(header.subarray(0, 16));
      const fileIndex = header.readUInt16BE(16);
      const streamIndex = header.readUInt16BE(18);
      const rest = header.subarray(20);

      const t = this.transfers.get(transferId);
      if (!t) { socket.destroy(); return; }
      const file = t.files[fileIndex];
      const ranges = byteRanges(file.size, PARALLEL_STREAMS);
      let pos = ranges[streamIndex].start;
      const fd = t.fds[fileIndex];

      const writeChunk = (buf) => {
        fs.writeSync(fd, buf, 0, buf.length, pos);
        pos += buf.length;
      };
      if (rest.length) writeChunk(rest);
      socket.on('data', writeChunk);
      socket.on('end', () => this._onStreamDone(transferId, fileIndex));
    });
  }

  _onStreamDone(transferId, fileIndex) {
    const t = this.transfers.get(transferId);
    if (!t) return;
    t._streamsDoneByFile = t._streamsDoneByFile || {};
    t._streamsDoneByFile[fileIndex] = (t._streamsDoneByFile[fileIndex] || 0) + 1;
    if (t._streamsDoneByFile[fileIndex] === PARALLEL_STREAMS) {
      fs.closeSync(t.fds[fileIndex]);
      this._maybeFinishFile(transferId);
    }
  }

  async _maybeFinishFile(transferId) {
    const t = this.transfers.get(transferId);
    if (!t) return;
    t._filesDone = (t._filesDone || 0) + 1;
    if (t._filesDone === t.files.length) {
      let ok = true;
      for (let i = 0; i < t.files.length; i++) {
        const hash = await sha256File(t.destPaths[i]);
        if (hash !== t.files[i].sha256) ok = false;
      }
      writeControlFrame(t.controlSocket, t.sessionKey, { type: 'complete', transferId, status: ok ? 'ok' : 'checksum_failed' });
      this.emit('complete', { transferId, status: ok ? 'ok' : 'checksum_failed', destPaths: t.destPaths });
      this.transfers.delete(transferId);
    }
  }

  stop() {
    if (this.controlServer) this.controlServer.close();
    for (const s of this.dataServers) s.close();
  }
}

class TransferClient {
  constructor(self, trustStore) {
    this.self = self;
    this.trustStore = trustStore;
  }

  /**
   * @param {object} peer - {deviceId, ip, transferPort}
   * @param {Array<{path:string,name:string,mime:string,isApp?:boolean,appLabel?:string,appPackage?:string}>} fileDescs
   * @returns {Promise<string>} resolves with status once receiver confirms 'complete'
   */
  async sendFiles(peer, fileDescs) {
    const sessionKey = this.trustStore.sessionKeyFor(peer.deviceId);
    if (!sessionKey) throw new Error('Cannot send: peer is not paired yet');

    const transferId = crypto.randomUUID();
    const files = await Promise.all(fileDescs.map(async (f) => ({
      name: f.name,
      size: fs.statSync(f.path).size,
      mime: f.mime,
      sha256: await sha256File(f.path),
      isApp: !!f.isApp,
      appLabel: f.appLabel,
      appPackage: f.appPackage,
    })));

    return new Promise((resolve, reject) => {
      const controlSocket = net.createConnection({ host: peer.ip, port: peer.transferPort }, () => {
        const helloPlain = Buffer.from(JSON.stringify({ type: 'hello', deviceId: this.self.deviceId }), 'utf8');
        const helloHeader = Buffer.alloc(4);
        helloHeader.writeUInt32BE(helloPlain.length, 0);
        controlSocket.write(Buffer.concat([helloHeader, helloPlain]));
        writeControlFrame(controlSocket, sessionKey, { type: 'offer', transferId, files });
      });

      controlFrameReader(controlSocket, sessionKey, async (msg) => {
        if (msg.type === 'accept' && msg.transferId === transferId) {
          await this._streamAllFiles(peer, transferId, fileDescs, files);
        }
        if (msg.type === 'complete' && msg.transferId === transferId) {
          controlSocket.end();
          resolve(msg.status);
        }
      });
      controlSocket.on('error', reject);
    });
  }

  async _streamAllFiles(peer, transferId, fileDescs, files) {
    for (let fileIndex = 0; fileIndex < fileDescs.length; fileIndex++) {
      const ranges = byteRanges(files[fileIndex].size, PARALLEL_STREAMS);
      await Promise.all(ranges.map((range, streamIndex) => new Promise((resolve, reject) => {
        const socket = net.createConnection({ host: peer.ip, port: peer.transferPort + streamIndex + 1 }, () => {
          const header = Buffer.alloc(20);
          uuidToBuf16(transferId).copy(header, 0);
          header.writeUInt16BE(fileIndex, 16);
          header.writeUInt16BE(streamIndex, 18);
          socket.write(header);
          const readStream = fs.createReadStream(fileDescs[fileIndex].path, { start: range.start, end: range.end });
          readStream.pipe(socket);
          readStream.on('end', () => socket.end());
        });
        socket.on('error', reject);
        socket.on('close', resolve);
      })));
    }
  }
}

module.exports = { TransferServer, TransferClient, PARALLEL_STREAMS };
